k256/src/schnorr.rs

```diff
@@ -0,0 +1,318 @@
+
+        /// Message digest
+        message: [u8; 32],
+
+        /// Expected signature
+        signature: [u8; 64],
+    }
+
+    /// BIP340 signing test vectors: index 0-3
+    const BIP340_SIGN_VECTORS: &[SignVector] = &[
+        // index 0
+        SignVector {
+            secret_key: hex!("0000000000000000000000000000000000000000000000000000000000000003"),
+            public_key: hex!("F9308A019258C31049344F85F89D5229B531C845836F99B08601F113BCE036F9"),
+            aux_rand: hex!("0000000000000000000000000000000000000000000000000000000000000000"),
+            message: hex!("0000000000000000000000000000000000000000000000000000000000000000"),
+            signature: hex!(
+                "E907831F80848D1069A5371B402410364BDF1C5F8307B0084C55F1CE2DCA8215
+                 25F66A4A85EA8B71E482A74F382D2CE5EBEEE8FDB2172F477DF4900D310536C0"
+            ),
+        },
+        // index 1
+        SignVector {
+            secret_key: hex!("B7E151628AED2A6ABF7158809CF4F3C762E7160F38B4DA56A784D9045190CFEF"),
+            public_key: hex!("DFF1D77F2A671C5F36183726DB2341BE58FEAE1DA2DECED843240F7B502BA659"),
+            aux_rand: hex!("0000000000000000000000000000000000000000000000000000000000000001"),
+            message: hex!("243F6A8885A308D313198A2E03707344A4093822299F31D0082EFA98EC4E6C89"),
+            signature: hex!(
+                "6896BD60EEAE296DB48A229FF71DFE071BDE413E6D43F917DC8DCF8C78DE3341
+                 8906D11AC976ABCCB20B091292BFF4EA897EFCB639EA871CFA95F6DE339E4B0A"
+            ),
+        },
+        // index 2
+        SignVector {
+            secret_key: hex!("C90FDAA22168C234C4C6628B80DC1CD129024E088A67CC74020BBEA63B14E5C9"),
+            public_key: hex!("DD308AFEC5777E13121FA72B9CC1B7CC0139715309B086C960E18FD969774EB8"),
+            aux_rand: hex!("C87AA53824B4D7AE2EB035A2B5BBBCCC080E76CDC6D1692C4B0B62D798E6D906"),
+            message: hex!("7E2D58D8B3BCDF1ABADEC7829054F90DDA9805AAB56C77333024B9D0A508B75C"),
+            signature: hex!(
+                "5831AAEED7B44BB74E5EAB94BA9D4294C49BCF2A60728D8B4C200F50DD313C1B
+                 AB745879A5AD954A72C45A91C3A51D3C7ADEA98D82F8481E0E1E03674A6F3FB7"
+            ),
+        },
+        // index 3
+        // TODO(tarcieri): failing; test vector notes: "test fails if msg is reduced modulo p or n"
+        // SignVector {
+        //     secret_key: hex!("0B432B2677937381AEF05BB02A66ECD012773062CF3FA2549E44F58ED2401710"),
+        //     public_key: hex!("25D1DFF95105F5253C4022F628A996AD3A0D95FBF21D468A1B33F8C160D8F517"),
+        //     aux_rand: hex!("FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF"),
+        //     message: hex!("FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF"),
+        //     signature: hex!(
+        //         "7EB0509757E246F19449885651611CB965ECC1A187DD51B64FDA1EDC9637D5EC
+        //          97582B9CB13DB3933705B32BA982AF5AF25FD78881EBB32771FC5922EFC66EA3"
+        //     ),
+        // },
```

Comment:
Though the other test vectors work, this one computes an incorrect signature:

```
thread 'schnorr::tests::bip340_sign_vectors' panicked at 'assertion failed: `(left == right)`
  left: `[126, 176, 80, 151, 87, 226, 70, 241, 148, 73, 136, 86, 81, 97, 28, 185, 101, 236, 193, 161, 135, 221, 81, 182, 79, 218, 30, 220, 150, 55, 213, 236, 151, 88, 43, 156, 177, 61, 179, 147, 55, 5, 179, 43, 169, 130, 175, 90, 242, 95, 215, 136, 129, 235, 179, 39, 113, 252, 89, 34, 239, 198, 110, 163]`,
 right: `[179, 77, 104, 140, 113, 27, 241, 243, 3, 27, 237, 173, 214, 221, 46, 61, 98, 13, 145, 199, 80, 58, 160, 96, 2, 171, 117, 128, 88, 175, 141, 194, 239, 225, 35, 117, 214, 130, 65, 221, 88, 38, 146, 1, 167, 11, 172, 54, 106, 122, 83, 94, 88, 208, 142, 255, 85, 40, 67, 58, 64, 34, 36, 59]`', k256/src/schnorr.rs:315:13
```

I'm uncertain why. As the comment notes, reducing the message mod n/p would cause a failure, however the current implementation only performs a reduction of the output of `tagged_hash` which seems orthogonal to what that's referring to.

Reply:
The notable property of this test vector per the code that generates it is that the secret key needs to be negated before signing:

https://github.com/bitcoin/bips/blob/274fa400d630ba757bec0c03b35ebe2345197108/bip-0340/test-vectors.py#L73